codegen/src/CMakeLists.txt

```diff
@@ -0,0 +1,62 @@
+# luci_codegen LIBRARY
+
+set(SOURCES Codegen.cpp
+        KernelBuilder.cpp
+        KernelBuilder.h
+        Scheduler.cpp
+        Scheduler.h
+        SubgraphContext.cpp
+        SubgraphContext.h
+        Utilities.cpp
+        Utilities.h
+        )
+
+set(WRAPPER_SOURCES ${LUCI_CODEGEN_SUPPORT_DIR}/WrapperLib.h)
+set(WRAPPER_LIB ${CMAKE_CURRENT_BINARY_DIR}/GeneratedWrapperLib.h)
+
+set(HEADERS ${LUCI_CODEGEN_INCLUDE_DIR}/Codegen.h ${LUCI_CODEGEN_INCLUDE_DIR}/CodegenOptions.h)
+
+add_library(luci_codegen SHARED ${SOURCES} ${HEADERS} ${WRAPPER_LIB})
+
+target_include_directories(luci_codegen PUBLIC "${LUCI_CODEGEN_INCLUDE_DIR}")
+target_include_directories(luci_codegen PRIVATE "${LUCI_CODEGEN_SOURCE_DIR}")
+target_include_directories(luci_codegen PRIVATE "${CMAKE_CURRENT_BINARY_DIR}")
+target_include_directories(luci_codegen PRIVATE ${Boost_INCLUDE_DIRS})
+
+target_link_libraries(luci_codegen PUBLIC luci_lang)
+target_link_libraries(luci_codegen PRIVATE nncc_common)
+target_link_libraries(luci_codegen PRIVATE Halide::Halide)
+target_link_libraries(luci_codegen PRIVATE flatbuffers)
+target_link_libraries(luci_codegen PRIVATE ${Boost_LIBRARIES})
+
+add_custom_command(OUTPUT ${WRAPPER_LIB}
+        COMMAND python3 ${LUCI_CODEGEN_SUPPORT_DIR}/embed.py ${LUCI_CODEGEN_SUPPORT_DIR}/WrapperLib.h ${CMAKE_CURRENT_BINARY_DIR}/GeneratedWrapperLib.h wrapper_lib_source
+        DEPENDS ${WRAPPER_SOURCES}
+        VERBATIM)
+
+install(TARGETS luci_codegen DESTINATION lib)
+
+# native_compiler EXECUTABLE
```

Comment:
`native_compiler` is like the `Native` ArchType? it's a bit ambigious for me... is this x86-64? (as we only work on x86-64 for compiler)

Reply:
"native" refers to machine code.
I want to highlight that this compiler should produce binary code, not byte code or source.

P.s. This is not entirely true (:disappointed: ), because utility generates binary kernels and some **source** code, that wraps Halide structures and provides backend agnostic interfaces.

Reply:
I see... then as I understand this is not aligned with `Native` ArchType ....
using `native` word here maybe confusing but I can't think of good name right now...

This maybe another scope but is the generated binary wrapped as Custom operator through `CircleCustom` IR?

Reply:
> This maybe another scope but is the generated binary wrapped as Custom operator through CircleCustom IR?

Yes, generated binary is an implementation of CircleCustom in IR. I'll try to describe it in more detail (I should work on wording in documentation more and fill "Ratinale" section already :smiley: )

CircleCustom node is a "declaration" of operation that should be executed. It stores name of generated kernel.

Along with circle file runtime (ONERT, luci-interpreter, etc.) should be provided with library that contains implementation of generated kernel(s).

Each kernel is accessed using three functions:
- Constructor (of type `ConstructorCompiledFunc`)
- Destructor (of type `DestructorCompiledFunc`)
- Implementation function (of type `CompiledFuncWrapper`)

Constructor and Destructor names are "create_<kernel_name>" and "free_<kernel_name>" for every generated kernel.
Implementation function name is not important.

Wrapper library I mentioned implements these three functions.

More detail

**Constructor**

This function takes input tensor sizes and creates structure of `ConfiguredCompiledFunc` type.
`ConfiguredCompiledFunc` structure contains pointer to implementation function and some context.
You can think of this structure as std::function object, just implemented in C.

This function probably should be called once during preparation phase of runtime. (or after arguments resize)

**Destructor**

This function takes `ConfiguredCompiledFunc` object and free context if any.

**Implementation function**

This function implements kernel. To invoke it you need to pass context and pointers to memory with arguments.

Like this:
```
  ConfiguredCompiledFunc kernel = create_generates_operator(ranks, dims);

  kernel.wrapper(kernel.configuration, pointers_to_buffers);

  free_generates_operator(kernel);
```

P.s.
I decided to introduce contructor/destructor concept, because it opens a lot of interesting opportunities:
1. Some computations could be done in constructor. For example I allocate memory for Halide shape structures and initialize some other fields.
2. Runtime can JIT compile kernel implementation itself! in that case there are no implementation function provided at all.
  This could be useful for autotune approach implemented in TVM,

P.p.s.
I just remembered that we already worked on Custom operations in onert. AFAIK we didn't use it so I forgot about this. Need to check if it is still working.